In [1]:
# Imports and pip installations (if needed)
!pip install pandas
import pandas as pd

# Part 1: Load the dataset

In [2]:
# Load the given datasets
numerical_data = pd.read_csv('chronic_kidney_disease_numerical.csv')
categorical_data = pd.read_csv('chronic_kidney_disease_categorical.csv')

# Print the data
print("Numerical Data:")
print(numerical_data.head())

print("\nCategorical Data:")
print(categorical_data.head())

Numerical Data:
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  8200.0   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN     NaN   

   rbcc  
0   6.4  
1   5.5  
2   5.7  
3   4.5  
4   NaN  

Categorical Data:
   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    norm

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: 

The chronic kidney disease (CKD) dataset from the UCI Machine Learning Repository includes data collected from various medical tests to predict whether a patient has chronic kidney disease. This dataset includes multiple features (input variables) and one label (output variable). Here’s a breakdown:

Features: These are the characteristics or attributes that can help in determining whether a person has chronic kidney disease. In this dataset, the features represent various health-related attributes of patients, such as blood pressure, glucose levels, and specific medical conditions.

Numerical Features: Blood Pressure (bp): The diastolic blood pressure in mm/Hg, Serum Creatinine (sc): The level of creatinine in the blood, measured in mg/dL., Hemoglobin (hemo): The hemoglobin level in g/dL.

Categorical Features: These are qualitative values that are typically represented by categories or classes. For example: Anemia (ane): This column indicates whether the patient has anemia or not (yes/no), Diabetes Mellitus (dm): This column indicates whether the patient has diabetes (yes/no), Hypertension (htn): This indicates if the patient has hypertension (yes/no).

Label: The label (output variable) is the outcome we are trying to predict, which in this case is whether a patient has chronic kidney disease (CKD) or not. It is typically represented as a categorical variable (e.g., "yes" or "no").


Dataset Composition: Numerical Features: These involve continuous values such as blood pressure, creatinine, and hemoglobin, Categorical Features: These are attributes that describe categories such as whether a patient has anemia, diabetes, or hypertension., The dataset includes both numerical and categorical features, which can be useful for training a machine learning model that aims to predict whether a person has CKD.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicate_rows_numerical = numerical_data[numerical_data.duplicated()]
print(f"Duplicate rows in numerical dataset:\n{duplicate_rows_numerical}")
# Drop these duplicate rows
numerical_data_cleaned = numerical_data.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicate_rows_categorical = categorical_data[categorical_data.duplicated()]
print(f"Duplicate rows in categorical dataset:\n{duplicate_rows_categorical}")

categorical_data_cleaned = categorical_data.drop_duplicates()

Duplicate rows in numerical dataset:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
107     872034  70.0   90.0  184.0   98.6  3.3  138.0  3.9   5.8   NaN   
149     721270  58.0   70.0  102.0   48.0  1.2  139.0  4.3  15.0  40.0   
171     240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
173     441170  79.0   80.0  111.0   44.0  1.2  146.0  3.6  16.3  40.0   
185     261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  43.0   
194     769512  38.0   60.0   91.0   36.0  0.7  135.0  3.7  14.0  46.0   
213     354223  64.0   70.0   97.0   27.0  0.7  145.0  4.8  13.8  49.0   
244     302906  59.0  100.0    NaN   96.0  6.4    NaN  NaN   6.6   NaN   
245     431871  57.0   60.0  132.0   18.0  1.1  150.0  4.7  15.4  42.0   
248     687021  52.0   80.0   99.0   25.0  0.8  135.0  3.7  15.0  52.0   
249     843449  30.0   80.0  131.0   38.0  1.0  147.0  3.8  14.1  45.0   
265     628271  65.0   90.0  270.0   39.0  2.0    NaN  NaN  12.0  36.0   
2

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No, it's likely that not all unique IDs are present in both datasets. This could be as we are dealing with medical or observational data, where not all patients or samples may have complete data across all variables (numerical and categorical). Missing data can occur for several reasons, such as incomplete tests or different recording procedures.

When performing a merge (specifically an "outer" merge that keeps all rows from both datasets), the rows that are missing from one dataset would have NaN (Not a Number) values in the columns corresponding to the dataset they are missing from. For example: If a patient’s numerical data is missing but their categorical data exists, the columns corresponding to numerical values (like blood pressure) would show NaN. Similarly, if categorical data is missing for some patients, the corresponding categorical columns (such as "hypertension") would show NaN values in those rows.

In [4]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_data = pd.merge(numerical_data_cleaned, categorical_data_cleaned, on='unique_id', how='inner')

#Print the combined dataset
print("Combined Dataset:")
print(combined_data.head())

Combined Dataset:
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   

       pc         pcc          ba  htn   dm cad appet   pe ane  Target  
0  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
1  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
2  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
3  normal  notpresent  notpresent  yes  yes  no  poor  yes  no     ckd  
4  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  

[5 rows x 25 columns]


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [5]:
# Calculate the percentage of rows that contain atleast one missing value

# Print %

missing_percentage = (combined_data.isnull().sum(axis=1) > 0).mean() * 100
print(f"Percentage of rows with missing values: {missing_percentage:.2f}%")

# Drop these rows from the dataset
cleaned_data = combined_data.dropna()
# Print the Dataset
print(cleaned_data)


Percentage of rows with missing values: 60.77%
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13      349892  59.0   70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
379     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
381     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
386     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0  40.0  1.7  125.0  3.5   8.3  23.0  ...   

           pc   

## Part 3.4: Sort the dataset according to the Labels

In [6]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_data = cleaned_data.sort_values(by='Target').reset_index(drop=True)

# Print the dataset
print(sorted_data)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [7]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_data = pd.get_dummies(sorted_data, columns=categorical_columns)

# Print the dataset
print(encoded_data)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [8]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for column in numerical_columns:
    mean = sorted_data[column].mean()
    std = sorted_data[column].std()
    sorted_data = sorted_data[(sorted_data[column] >= mean - 3 * std) & (sorted_data[column] <= mean + 3 * std)]

# Print the dataset
print (sorted_data)


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
6       546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [9]:
# Normalize the all Numerical Attributes in the dataset.
from sklearn.preprocessing import MinMaxScaler

# Apply MinMaxScaler
scaler = MinMaxScaler()
sorted_data[numerical_columns] = scaler.fit_transform(sorted_data[numerical_columns])

# Print the dataset
# 133 x 36

print(sorted_data[numerical_columns])


          age    bp       bgr        bu        sc       sod       pot  \
0    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
1    0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
3    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
4    0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.793103   
6    0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.743243  0.25  0.214592  0.124183  0.037037  0.566667  0.206897   
149  0.418919  0.00  0.167382  0.241830  0.086420  0.700000  0.689655   
150  0.689189  0.50  0.210300  0.045752  0.098765  0.500000  0.620690   
151  0.472973  0.25  0.236052  0.183007  0.024691  0.666667  0.724138   
152  0.662162  0.50  0.145923  0.117647  0.061728  0.733333  0.655172   

         hemo       pcv      wbcc      rbcc  
0    0.059406  0.000000  0.653226  0.257143  
1    0.148515  0.225806  0.2177

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: 

Reasons for Dropping unique_id: Unique identifiers do not carry any predictive value. They are arbitrary and have no relationship to the target variable (such as the diagnosis or condition being predicted, e.g., CKD or notCKD). Since unique identifiers are random or sequential, they can't be used in any meaningful calculations (such as averages, summations, or correlations with the target label). Keeping them in the model might introduce noise rather than help the model learn meaningful patterns. Including a unique identifier in a model can result in the model memorizing rather than learning patterns from data. For instance, if a unique ID is included in training, the model might falsely "learn" that a specific ID is associated with a particular outcome, leading to overfitting.

Hence we would drop the unique_id or patient_id column because it does not contribute to predictions or calculations, and its inclusion could lead to noise in the model rather than meaningful insights. 

In [10]:
#Remove that column
merged_data_cleaned = sorted_data.drop(columns=['unique_id'])  # Replace 'patient_id' with the actual column you want to drop

# Print the dataset
print(merged_data_cleaned)

          age    bp       bgr        bu        sc       sod       pot  \
0    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
1    0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
3    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
4    0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.793103   
6    0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.743243  0.25  0.214592  0.124183  0.037037  0.566667  0.206897   
149  0.418919  0.00  0.167382  0.241830  0.086420  0.700000  0.689655   
150  0.689189  0.50  0.210300  0.045752  0.098765  0.500000  0.620690   
151  0.472973  0.25  0.236052  0.183007  0.024691  0.666667  0.724138   
152  0.662162  0.50  0.145923  0.117647  0.061728  0.733333  0.655172   

         hemo       pcv      wbcc  ...        pc         pcc          ba  htn  \
0    0.059406  0.000000  0.653226  ...  ab

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [11]:
# Export the dataframe to a new csv file
merged_data_cleaned.to_csv('cleaned_data.csv', index=False)

# Export the dataframe to a new json file

merged_data_cleaned.to_json('cleaned_data.json')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here

|    |    age    |   bp   |    bgr    |    bu    |    sc    |    sod   |    pot   |    hemo    |    pcv    |    wbcc   |    pc      |    pcc      |    ba        |    htn   |    dm    |    cad   |    appet   |    pe    |    ane   |    Target   |
|----|-----------|--------|-----------|----------|----------|----------|----------|------------|-----------|-----------|------------|-------------|--------------|----------|----------|----------|------------|----------|----------|-------------|
|  0 |  0.202703 |  0.75  |  0.158798 |  0.196078 |  0.160494 |  0.166667 |  0.206897 |  0.059406  |  0.000000 |  0.653226 |  abnormal  |  present    |  present     |  no      |  no      |  no      |  good      |  no      |  yes     |  ckd        |
|  1 |  0.905405 |  1.00  |  0.965665 |  0.522876 |  0.641975 |  0.666667 |  0.000000 |  0.148515  |  0.225806 |  0.217742 |  abnormal  |  present    |  notpresent  |  yes     |  yes     |  yes     |  poor      |  no      |  no      |  ckd        |
|  3 |  0.743243 |  0.50  |  0.442060 |  0.901961 |  0.432099 |  0.500000 |  0.793103 |  0.000000  |  0.032258 |  0.395161 |  abnormal  |  notpresent |  notpresent  |  yes     |  yes     |  yes     |  poor      |  yes     |  yes     |  ckd        |
|  4 |  0.729730 |  0.75  |  0.150215 |  0.281046 |  0.234568 |  0.533333 |  0.793103 |  0.336634  |  0.322581 |  0.500000 |  normal    |  notpresent |  notpresent  |  no      |  no      |  no      |  good      |  no      |  no      |  ckd        |
|  6 |  0.540541 |  0.00  |  0.399142 |  0.535948 |  0.358025 |  0.700000 |  0.379310 |  0.207921  |  0.161290 |  0.830645 |  normal    |  notpresent |  notpresent  |  yes     |  yes     |  no      |  good      |  no      |  no      |  ckd        |
|  8 |  0.675676 |  0.75  |  0.253219 |  0.633987 |  0.777778 |  0.366667 |  0.655172 |  0.138614  |  0.193548 |  0.169355 |  abnormal  |  notpresent |  notpresent  |  yes     |  no      |  no      |  good      |  no      |  no      |  ckd        |
| 10 |  0.716216 |  0.50  |  1.000000 |  0.163399 |  0.111111 |  0.066667 |  0.206897 |  0.267327  |  0.387097 |  0.532258 |  normal    |  notpresent |  notpresent  |  no      |  yes     |  no      |  poor      |  no      |  no      |  ckd        |
| 13 |  0.729730 |  0.00  |  0.935622 |  0.169935 |  0.160494 |  0.333333 |  0.034483 |  0.019802  |  0.064516 |  0.879032 |  abnormal  |  present    |  notpresent  |  yes     |  no      |  no      |  poor      |  no      |  yes     |  ckd        |
| 17 |  0.000000 |  0.00  |  0.103004 |  0.372549 |  0.074074 |  0.500000 |  0.689655 |  0.217822  |  0.225806 |  1.000000 |  abnormal  |  notpresent |  present     |  no      |  no      |  no      |  poor      |  no      |  no      |  ckd        |
| 19 |  0.878378 |  0.00  |  0.206009 |  0.751634 |  0.604938 |  0.533333 |  0.689655 |  0.366337  |  0.387097 |  0.879032 |  normal    |  notpresent |  notpresent  |  yes     |  yes     |  no      |  poor      |  yes     |  no      |  ckd        |
| 20 |  0.851351 |  0.25  |  0.618026 |  0.562092 |  0.728395 |  0.000000 |  0.344828 |  0.168317  |  0.161290 |  0.580645 |  abnormal  |  present    |  present     |  yes     |  yes     |  yes     |  good      |  yes     |  yes     |  ckd        |
| 21 |  0.878378 |  0.25  |  0.639485 |  0.470588 |  0.395062 |  0.433333 |  0.517241 |  0.267327  |  0.322581 |  0.104839 |  abnormal  |  present    |  present     |  yes     |  yes     |  yes     |  good      |  no      |  no      |  ckd        |
| 22 |  0.783784 |  0.00  |  0.725322 |  0.313725 |  0.481481 |  0.566667 |  0.862069 |  0.178218  |  0.193548 |  0.258065 |  abnormal  |  notpresent |  present     |  yes     |  yes     |  no      |  poor      |  yes     |  no      |  ckd        |
| 24 |  0.662162 |  0.50  |  0.618026 |  0.411765 |  0.432099 |  0.566667 |  0.689655 |  0.316832  |  0.354839 |  0.250000 |  abnormal  |  present    |  present     |  yes     |  yes     |  no      |  good      |  yes     |  no      |  ckd        |
| 25 |  0.770270 |  1.00  |  0.901288 |  0.163399 |  0.345679 |  0.766667 |  0.206897 |  0.524752  |  0.548387 |  0.443548 |  normal    |  notpresent |  present     |  yes     |  no      |  yes     |  good      |  no      |  no      |  ckd        |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [12]:
# Dictionary of old column names to new column names
column_renames = {
    'bp': 'blood_pressure',
    'sg': 'specific_gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red_blood_cells',
    'pc': 'pus_cell',
    'pcc': 'pus_cell_clumps',
    'ba': 'bacteria',
    'bgr': 'blood_glucose_random',
    'bu': 'blood_urea',
    'sc': 'serum_creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed_cell_volume',
    'wbcc': 'white_blood_cell_count',
    'rbcc': 'red_blood_cell_count',
    'htn': 'hypertension',
    'dm': 'diabetes_mellitus',
    'cad': 'coronary_artery_disease',
    'appet': 'appetite',
    'pe': 'pedal_edema',
    'ane': 'anemia',
    'classification': 'target'
}
# Code to rename all the columns in the dataset
merged_data_cleaned.rename(columns=column_renames, inplace=True)
print(merged_data_cleaned.head())

        age  blood_pressure  blood_glucose_random  blood_urea  \
0  0.202703            0.75              0.158798    0.196078   
1  0.905405            1.00              0.965665    0.522876   
3  0.743243            0.50              0.442060    0.901961   
4  0.729730            0.75              0.150215    0.281046   
6  0.540541            0.00              0.399142    0.535948   

   serum_creatinine    sodium  potassium  hemoglobin  packed_cell_volume  \
0          0.160494  0.166667   0.206897    0.059406            0.000000   
1          0.641975  0.666667   0.000000    0.148515            0.225806   
3          0.432099  0.500000   0.793103    0.000000            0.032258   
4          0.234568  0.533333   0.793103    0.336634            0.322581   
6          0.358025  0.700000   0.379310    0.207921            0.161290   

   white_blood_cell_count  ...  pus_cell  pus_cell_clumps    bacteria  \
0                0.653226  ...  abnormal          present     present   
1     

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [13]:
# Converted SQL to Pandas code
target_count = merged_data_cleaned.groupby('Target').size().reset_index(name='count')
print(target_count)

   Target  count
0     ckd     22
1  notckd    111


*Describe what the above code does here*



What the Pandas Code Does:
-groupby('Target'): Groups the rows in the dataset by the unique values in the 'Target' column (just like the GROUP BY in SQL).
-.size(): Counts the number of rows for each group (like the COUNT(*) in SQL).
-reset_index(name='count'): Resets the index and names the resulting count column as 'count'.
-After running this, you would get a DataFrame that shows how many times each unique Target value appears in the dataset, mimicking the SQL query result.